### DISCUSSION TITLE
优化delegate之后，C++无法Clear掉TS注册的回调

### DISCUSSION THREAD

**Discussion by watsonsong** [Tencent/puerts#1575] [Q&A] [answered]
发现在这个[优化](https://github.com/Tencent/puerts/issues/1510)之后，如果C++层面对一个Delegate执行Unbind或者Clear操作，TS注册的回调无法被正确的清理。

**Comment by chexiongsheng** [comment]
后面有commit解决了

**Reply by watsonsong** [reply]
我更新到最新的master分支依然会有这个情况，就是C++中对一个DYNAMIC_DELEGATE执行了Unbind之后，后续的TS再注册这个delegate回调总是触发不了，似乎delegate的状态有一些问题。
常见的是对象池回收的时候，希望直接清理掉整个delegate，再次拿出来用的时候Bind的TS回调没法再触发了。

似乎是在C++层面Unbind之后，`FJsEnvImpl::AddToDelegate`里面，从`DelegateMap`中找到的`Iter->second.Proxy.IsValid()`依然判定是有效的。但其实在C++层面因为经历过Unbind，它的DelegatePtr里面已经没有之前加入的DelegateProxy了。但是DelegateMap中还存在，这样TS还在往一个老的DelegateProxy中加入。

具体到我碰到的对象池的案例，如果我在还回池之前，由JS去调用Unbind就可以避免这个问题。但很多时候不可避免的还是需要支持C++的Unbind清理一个回调。特别是使用C++的一些独立模块，不依赖JS的情况下要能正常工作。

**Reply by chexiongsheng** [reply]
能搞个重现demo么？

**Reply by chexiongsheng** [reply]
cleardelegate的时候有释放DelegateProxy： https://github.com/Tencent/puerts/blob/master/unreal/Puerts/Source/JsEnv/Private/JsEnvImpl.cpp#L2605
我看你提问题之后也有针对cleardelegate的提交，你更新试了么？

**Reply by BlurryLight** [reply]
是会有这个问题，因为这个判断太弱了，至少要加入判断这个Delegate是否`IsBound()`的检查。因为一个DynamicDelegate可能在C++模块执行Reset操作(不经过FJsEnvImpl::ClearDelegate)，puerts是没感知的。 正确的处理需要在这一行判断delegate是否已经被外部代码解绑了，如果被外部代码解绑了需要重新申请一个新的DelegateProxy绑上去。

https://github.com/Tencent/puerts/blob/1468340b87d0ed768cfb908d4b4ff84da2cd55bb/unreal/Puerts/Source/JsEnv/Private/JsEnvImpl.cpp#L2358

有空我贴一个最简复现代码在这里

**Reply by chexiongsheng** [reply, marked as answer]
我这里重现了，也修复了，试试合入这个commit：https://github.com/Tencent/puerts/commit/3500dff0b328443077a59a8ce05d701d61b133d2 ， @watsonsong  ， @BlurryLight
